mono/utils/mono-error-internals.h

```diff
@@ -284,4 +284,80 @@ mono_error_set_specific (MonoError *error, int error_code, const char *missing_m
 void
 mono_error_set_first_argument (MonoError *oerror, const char *first_argument);
 
+// Provide single instruction inlinable forms of GetLastError and SetLastError.
+// Nobody should worry about their cost then.
+// Offsets are from Wikipedia.
+// Runtime disassembly could assert correctness.
+
+// FIXME include this file more so the replacement takes affect.
+
+#if HOST_WIN32 && (HOST_X86 || HOST_AMD64) && (_MSC_VER || __GNUC__)
+
+#include <windows.h>
+
+// Naming violation so can search disassembly for GetLastError / SetLastError.
+#define GetLastError mono_GetLastError
+#define SetLastError mono_SetLastError
+
+static inline
+unsigned long
+__stdcall
+GetLastError (void)
+{
+#if __GNUC__
+    unsigned long err;
+
+#if HOST_X86
+    __asm__ ("movl %%fs:0x34, %0" : "=r" (err));
+#elif HOST_AMD64
+    __asm__ ("movl %%gs:0x68, %0" : "=r" (err));
+#else
+#error unknown architecture
+#endif
+    return err;
+
+#elif _MSC_VER
+#if HOST_X86
+    return __readfsdword (0x34);
+#elif HOST_AMD64
+    return __readgsdword (0x68);
+#else
+#error unknown architecture
+#endif
+
+#else
+#error unknown compiler
+#endif
+}
+
+static inline
+void
+__stdcall
+SetLastError (unsigned long err)
+{
+#if __GNUC__
+#if HOST_X86
+    __asm__ ("movl %0, %%fs:0x34" : : "r" (err));
```

Comment:
I think gcc has the intrinsics actually, so can remove `#if`.

Reply:
Or at least macros with this interface that generate the same assembly.

Reply:
I think it may support it but I will rather check that it works on both x86/x64.